lib/util.c

```diff
@@ -2187,9 +2194,332 @@ EXPORTED void xsyslog_fn(int priority, const char *description,
+
+    /*
+     * Now handle the remaining arguments, which should come in threes:
+     *   key, format, value
+     *
+     * value will be buf_printf'ed with the format,
+     * and then the result will be escaped (if necessary)
+     *
+     * finally, the resulting string "key=escaped-formatted-value"
+     * will be appended to the log buffer
+     */
+    while ((key = va_arg(ap, const char *))) {
+        // special case: __sep__ means reconfigure the separator
+        if (!strcmp(key, "__sep__")) {
+            sep = va_arg(ap, const char *);
+        }
+        else {
+            const char *fmt = va_arg(ap, const char *);
+            va_list val_ap;
+
+            va_copy(val_ap, ap);
+            buf_reset(&val);
+            buf_vprintf(&val, fmt, val_ap);
+            va_end(val_ap);
+
+            buf_printf(&log, "%s%s=%s",
+                       sep, key, xsyslog_ev_escape_value(&val));
+
+            /* Unfortunately neither ap nor val_ap is now pointing to the next 'key' argument.
+             * .. val_ap has been passed to vsprintf, which renders is unfit for further use.
+             .. ap is still pointing to 'val'.
+
+             This function understands `fmt` well enough that it can discard the
+             correct `val` argument and advance `ap` to point to the next `key`.
+            */
+            xsyslog_ev_fmt_skip_va(fmt, ap);
+
+        }
+    }
+
+    buf_free(&val);
+
+    return buf_release(&log);
+}
+
+static void xsyslog_ev_fmt_skip_va(const char* fmt, va_list ap)
+{
+    switch (xsyslog_ev_guess_printf_escape(fmt)) {
+    case 'd': (void) va_arg(ap, int);      break;
+    case 'l': (void) va_arg(ap, long);     break;
+    case 'f': (void) va_arg(ap, double);   break;
+    case 's': (void) va_arg(ap, char *); break;
+    case 'p': (void) va_arg(ap, void *); break;
+    case 0:
+        // we could handle this by skipping the va_arg call entirely, which would allow the user to write
+        //     xsyslog_ev(..., key, "foo", ...)
+        // in place of
+        //     xsyslog_ev(..., key, "foo", DUMMY, ...)
+        // but that seems more likely to be confusing than otherwise
+    default:
+        // This seems extreme at first but the alternative is to leave ap pointing to the wrong argument,
+        // and that is already undefined behavior, likely a segfault, so this is really the best we can do
+        abort();
+    }
+}
+
+/* Quick-and-dirty parser for SINGLE, COMMON printf-type escape strings
+ *
+ * return -1 if string can't be recognized
+ * return 0 if string contains no %-escapes
+ * otherwise return:
+ *   d   if the expected argument is an int
+ *   l   if the expected argument is a long
+ *   f   if the expected argument is a double
+ *   s   if the expected argument is a (char *)
+ *   p   if the expected argument is a (void *)
+ *   -1  if it's something else
+ */
+static char xsyslog_ev_guess_printf_escape(const char *s) {
```

Comment:
This should probably return `int` rather than `char`.  It's fine for it to be logically a single character still, but it's usual in C to use the `int` type for single characters, for reasons I don't quite remember, but which might be related to the implicit type promotion rules and/or the fact that the signedness of `char` is platform-dependent.

See for example the `int c` argument to `char *strchr(const char *s, int c);`